tortoisebeacon/weakcoin/weak_coin.go

```diff
@@ -4,290 +4,303 @@ import (
+	for unit := uint64(0); unit < allowed; unit++ {
+		proposal := wc.encodeProposal(epoch, round, unit)
+		signature := wc.signer.Sign(proposal)
+		if wc.exceedsThreshold(signature) {
+			continue
+		}
+		if smallest == nil || bytes.Compare(signature, smallest) == -1 {
+			message := Message{
+				Epoch:     epoch,
+				Round:     round,
+				Unit:      unit,
+				Signature: signature,
+			}
+			msg, err := types.InterfaceToBytes(message)
+			if err != nil {
+				wc.logger.Panic("can't serialize weak coin message", log.Err(err))
+			}
+
+			broadcast = msg
+			smallest = signature
+		}
 	}
-
-	serializedMessage, err := types.InterfaceToBytes(message)
-	if err != nil {
-		return fmt.Errorf("serialize weak coin message: %w", err)
+	// nothing to send is valid if all proposals are exceeding threshold
+	if broadcast == nil {
+		return nil
 	}
 
-	if err := wc.net.Broadcast(ctx, GossipProtocol, serializedMessage); err != nil {
-		return fmt.Errorf("broadcast weak coin message: %w", err)
+	if err := wc.net.Broadcast(ctx, GossipProtocol, broadcast); err != nil {
+		return fmt.Errorf("failed to broadcast weak coin message: %w", err)
 	}
 
-	wc.Log.With().Info("Published proposal",
-		log.Uint64("epoch_id", uint64(epoch)),
+	wc.logger.With().Info("published proposal",
+		log.Uint32("epoch_id", uint32(epoch)),
 		log.Uint64("round_id", uint64(round)),
-		log.String("proposal", types.BytesToHash(signedProposal).ShortString()))
+		log.String("proposal", types.BytesToHash(smallest).ShortString()))
 
+	wc.mu.Lock()
+	defer wc.mu.Unlock()
+	if wc.epoch != epoch || wc.round != round {
+		// another epoch/round was started concurrently
+		return nil
+	}
+	wc.updateSmallest(smallest)
 	return nil
 }
 
-func (wc *weakCoin) proposalExceedsThreshold(proposal []byte) bool {
-	proposalInt := new(big.Int).SetBytes(proposal)
-
-	return proposalInt.Cmp(wc.threshold) == 1
-}
-
-func (wc *weakCoin) generateProposal(epoch types.EpochID, round types.RoundID) ([]byte, error) {
-	proposal := bytes.Buffer{}
-
-	proposal.WriteString(proposalPrefix)
-
-	epochBuf := make([]byte, 8)
-	binary.LittleEndian.PutUint64(epochBuf, uint64(epoch))
-	if _, err := proposal.Write(epochBuf); err != nil {
-		return nil, fmt.Errorf("write epoch to buffer: %w", err)
-	}
-
-	roundBuf := make([]byte, 8)
-	binary.LittleEndian.PutUint64(roundBuf, uint64(round))
-	if _, err := proposal.Write(roundBuf); err != nil {
-		return nil, fmt.Errorf("write round to buffer: %w", err)
+// CompleteRound computes coinflip based on proposals received in this round.
+// After it is called new proposals for this round won't be accepted.
+func (wc *WeakCoin) CompleteRound() {
+	wc.mu.Lock()
+	defer wc.mu.Unlock()
+	if wc.smallestProposal == nil {
+		wc.logger.With().Error("completed round without valid proposals",
+			log.Uint32("epoch_id", uint32(wc.epoch)),
+			log.Uint64("round_id", uint64(wc.round)),
+		)
+		return
 	}
-
-	return proposal.Bytes(), nil
+	coinflip := wc.smallestProposal[0]&1 == 1
```

Comment:
The weak coin should be the least significant bit of the signature. Why do you assume that a signature is always stored as a little-endian number?

Reply:
i am not sure that i understand the question. signature doesn't depend on the architecture in any way. the representation is always the same

Reply:
i guess your question was about how library that produces signature treats them (as big endian or little endian). i haven't thought about this. but found https://datatracker.ietf.org/doc/html/draft-josefsson-eddsa-ed25519-03 Section 
5.2.  Encoding

>  All values are coded as octet strings, and integers are coded using
   little endian convention.  I.e., a 32-octet string h h[0],...h[31]
   represents the integer h[0] + 2^8 h[1] + ... + 2^248 h[31].

Reply:
it may become invalid if we will switch to another library. but idk if it is worth it to add it to the interface of the signer/verifier

Reply:
I meant that to extract an LSB from bytes, we need to get `wc.smallestProposal[0]` if a proposal is stored as a little-endian number and `wc.smallestProposal[len(wc.smallestProposal)-1]` if the proposal is stored as a big-endian number. So now we assume that a proposal is a little-endian number. Of course, if we get not LSB but some other byte consistently on all nodes, the algorithm would work. But I'd add a comment for this.

Reply:
> i guess your question was about how library that produces signature treats them (as big endian or little endian)

Yes, that's what I meant. Thanks for checking. So it would work fine. Could you please add a comment with something like this?

```suggestion
	// ED25519 assumes that byte order is little-endian. For another algorithm, the endianness might differ.
	coinflip := wc.smallestProposal[0]&1 == 1
```